Initial implementation

pavolloffay opened the pull request:
Import from old repository with an initial implementation.

How does this work? 
* Span is started and finished in servlet filter
* if span created in filter is finished followsFrom span is created. It is finished in `afterCompletion` for the same handler which was passed to `preHandle`. Context of this span overrides context of span created in servlet filter (so created spans in error controllers will be children of it).
* Decorator adds log when given controller has been invoked.

Test coverage:
- [x] sync
- [x] async 
- [x] forward
- [x] redirect
- [x] view
- [x] controller view
- [x] exception in filter
- [x] exception in controller
- [x] mapped exception
- [x] local span
- [x] no url mapping
- [x] secured endpoints

These tests are run for embedded boot deployed on tomcat and also for MVC deployed on jetty.

This depends on:
* https://github.com/opentracing-contrib/java-web-servlet-filter/pull/8
* https://github.com/opentracing/opentracing-java/pull/98 with new patch version

pavolloffay commented:
@mabn repository has been moved here. Thanks for your previous review. If you would like to look at this again it would be great :+1: .

mabn commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/SpanDecorator.java:
```diff
@@ -0,0 +1,59 @@
+        @Override
+        public void onAfterCompletion(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse,
+                                      Object handler, Exception ex, Span span) {
+            Tags.HTTP_STATUS.set(span, httpServletResponse.getStatus());
```
I'd also add an error tag (see https://github.com/opentracing/specification/blob/master/data_conventions.yaml#L16)
I'm not sure when exactly though. Probably testing for `ex != null` and ignoring http status code would be fine.

pavolloffay replied:
Why would you ignore `http.status_code` tag? IIRC server can respond with different status codes for different exceptions.

I was also thinking about `error` tag when exception was added to `onAfterCompletion`. Finally it's there :+1:

mabn replied:
I meant not to base the decision about adding an error ta on the status code but only on the presence of the exception. This way you avoid having to decide if 404 is an error or just a nice REST response for "record not found"

mabn commented:
Looks nice, I especially like those decorators.

One thing I'm concerned about are requests which won't hit the interceptor. I believe that for example spring security filter chain is executed before interceptors. Same thing would be true to request which do not match any request mapping or cause exceptions in some server filter. Ideally I'd like a servlet filter that can be registered as the first one in the filter chain which would cooperate with the interceptor - servlet filter would open a span and fill some basic informations (http method, uri, exception if thrown) and the interceptor could enrich it with additional details - like the exception thrown from the controller or the details of the handler that handled the request.

Although it is already possible to achieve with some custom code thanks to TracingDecision having access to request attributes.

pavolloffay commented:
@mabn Thanks for the review! It is very helpful :+1: 

Issues highlighted in your comment:
1. requests which do not match any URL mapping. Tested interceptor is invoked.
2. Exceptions in filter: Interceptor is invoked however exception in `afterCompletion(..ex..)` is null? But there is a solution for it: request attribute with name: (`org.springframework.web.util.WebUtils.ERROR_EXCEPTION_ATTRIBUTE`) .
3. Exception in controller: `afterCompletion` is called twice first with exception and status code 200 and second time with an exception null but exception is in request attributes (see point 2 exception in filters) and correct status code e.g. 500. Is it guaranteed that on an exception in controller `afterCompletion` is called twice? 
4. Spring security: interceptor is invoked for secured endpoints.


This currently depends on https://github.com/opentracing-contrib/java-spring-web, `HttpServletHeadersExtractAdapter(httpServletRequest))`

mabn commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/TracingHandlerInterceptor.java:
```diff
@@ -0,0 +1,103 @@
+            }
+            spanWrapper = new SpanWrapper(serverSpan, true);
+            httpServletRequest.setAttribute(ACTIVE_SPAN, spanWrapper);
+        }else if (spanWrapper == null && httpServletRequest.getAttribute(ACTIVE_SPAN) == null &&
```
`spanWrapper == null && httpServletRequest.getAttribute(ACTIVE_SPAN) == null`

I guess one of above is sufficient as spanWrapper == httpServletRequest.getAttribute(ACTIVE_SPAN)

pavolloffay commented:
@gbrown @jpkrohling could you please review? 

review notes: 
on exceptions in controller `afterCompletion` the first call has a wrong status code. The second call is ignored because span has been finished. We should not probably add status code if there is an exception

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/SpanDecorator.java:
```diff
@@ -0,0 +1,107 @@
+            if (httpServletRequest.getAttribute(WebUtils.ERROR_REQUEST_URI_ATTRIBUTE) != null) {
+                try {
+                    URL url = new URL(httpServletRequest.getRequestURL().toString());
+                    url.getProtocol();
```
This code is lost here :)

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/SpanDecorator.java:
```diff
@@ -0,0 +1,107 @@
+                try {
+                    URL url = new URL(httpServletRequest.getRequestURL().toString());
+                    url.getProtocol();
+                    String errorUrl = url.getProtocol() + "://" + url.getHost() + ":" + url.getPort() +
```
Would be more readable with `String.format("%s://%s:%s%s", protocol, host, port, errorRequestUriAttribute)`

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/SpanDecorator.java:
```diff
@@ -0,0 +1,107 @@
+                    Tags.HTTP_URL.set(span, errorUrl);
+                } catch (MalformedURLException e) {
+                    log.severe("failed to construct error URL");
+                    e.printStackTrace();
```
Instead of printing the stack trace, see if you can pass the exception to the log.severe/log.error . This way, the system administrator can shut it up in case it's too noisy. If you do a `e.printStackTrace`, there's no way to disable just this one.

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/SpanDecorator.java:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        private Map<String, String> logsForException(Throwable throwable) {
+            Map<String, String> errorLog = new HashMap<>();
```
You know for sure you will have only 3 entries on the map. Use the constructor to tell it to allocate only 3 positions, instead of the 16 that it does by default.

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/SpanDecorator.java:
```diff
@@ -0,0 +1,107 @@
+                    throwable.getMessage();
+            errorLog.put("message", exMessage);
+
+            StringWriter sw = new StringWriter();
```
So, you are adding `event`, `message` and `stack trace` to the span. Probably not something for now, but you might want to add a todo item to allow customization of this. I can think of several reasons not to log the entire stack trace to a span :)

pavolloffay replied:
Users can implement their own span decorator. I understand your point.

yurishkuro replied:
Personally, I would do this differently. I would log the following fields only:
  * event: "exception"
  * exception: throwable

and let the tracer implementation deal with exact representation of the exception / stack trace.

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/SpanWrapper.java:
```diff
@@ -0,0 +1,45 @@
+     */
+    public synchronized void finish() {
+        if (!finished.get()) {
+            finished.set(true);
```
You can probably get rid of the synchronized on the method if you use the `getAndSet` from the `AtomicBoolean`. Something like this:

```java
public void finish() {
    boolean wasFinished = finished.getAndSet(true); // this returns the previous value of finished and is atomic
    if (!wasFinished && !spanFromFilter) {
      span.finish();
    }
}
```

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/TracingHandlerInterceptor.java:
```diff
@@ -0,0 +1,103 @@
+                decorator.onPreHandle(httpServletRequest, handler, serverSpan);
+            }
+            spanWrapper = new SpanWrapper(serverSpan, true);
+        } else if (httpServletRequest.getAttribute(ACTIVE_SPAN) == null &&
```
Isn't `httpServletRequest.getAttribute(ACTIVE_SPAN)` already stored locally as `serverSpan`?

pavolloffay replied:
Yes here it has to be `ACTIVE_SPAN_WRAPPER`

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/TracingHandlerInterceptor.java:
```diff
@@ -0,0 +1,103 @@
+
+            // when exception is thrown in controller this is method is called twice, therefore using
+            // spanWrapper (span#finish()) is not idempotent by specification.
+            spanWrapper = new SpanWrapper(serverSpan, false);
```
Oh, *that's* why you use a Wrapper :) I was wondering about it until this point. Perhaps it's worth adding a javadoc to SpanWrapper about the motivation for such a class?

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/TracingHandlerInterceptor.java:
```diff
@@ -0,0 +1,103 @@
+            // when exception is thrown in controller this is method is called twice, therefore using
+            // spanWrapper (span#finish()) is not idempotent by specification.
+            spanWrapper = new SpanWrapper(serverSpan, false);
+            httpServletRequest.setAttribute(ACTIVE_SPAN, spanWrapper);
```
Is this correct? You set the `ACTIVE_SPAN` as the wrapper, but you seem to expect the `ACTIVE_SPAN` to be a `Span` (line 53)? Didn't you mean `ACTIVE_SPAN_WRAPPER`?

pavolloffay replied:
This is wrong.. I moved setting ACTIVE_SPAN_WRAPPER out of if-else. And forgot to remove this. 

Note that `ACTIVE_SPAN_WRAPPER` is used internally by interceptor and `ACTIVE_SPAN` is 'public' meant to be used by client code.

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/TracingHandlerInterceptor.java:
```diff
@@ -0,0 +1,103 @@
+    public void afterCompletion(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse,
+                                Object handler, Exception ex) throws Exception {
+
+        SpanWrapper spanWrapper = (SpanWrapper) httpServletRequest.getAttribute(ACTIVE_SPAN);
```
Isn't there a possibility of a `ACTIVE_SPAN` to be a `Span` instead of `SpanWrapper`? If so, you'd get a ClassCastException here.

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/test/java/io/opentracing/contrib/spring/web/servlet/AppConfiguration.java:
```diff
@@ -0,0 +1,150 @@
+        return new ExceptionFilter();
+    }
+
+//    @Bean
```
Left over?

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/test/java/io/opentracing/contrib/spring/web/servlet/AppConfiguration.java:
```diff
@@ -0,0 +1,150 @@
+    public FilterRegistrationBean tracingFilterRegistration() {
+        FilterRegistrationBean registrationBean = new FilterRegistrationBean();
+        TracingFilter tracingFilter = new TracingFilter(mockTracer, Arrays.asList(
+                io.opentracing.contrib.web.servlet.filter.SpanDecorator.STANDARD_TAGS),
```
`static import` those constants?

pavolloffay replied:
This is a left over when I was testing integration with servlet filter. The whole block is removed.

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/test/java/io/opentracing/contrib/spring/web/servlet/TracingHandlerInterceptorTest.java:
```diff
@@ -0,0 +1,256 @@
+            HttpEntity<String> entity = new HttpEntity<>(headers);
+
+            restTemplate.exchange("/sync", HttpMethod.GET, entity, String.class);
+            Awaitility.await().until(reportedSpansSize(), IsEqual.equalTo(1));
```
Interesting, I did something similar for APM, as I didn't find any libs that would do this. Remind me to replace our custom made "Wait#until()" by this :)

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/test/java/io/opentracing/contrib/spring/web/servlet/TracingHandlerInterceptorTest.java:
```diff
@@ -0,0 +1,256 @@
+
+    @Test
+    public void testAsync() throws Exception {
+        {
```
Why this anonymous block inside the method?

pavolloffay replied:
Actually I like this I can clearly see that nothing from inside a block is propagated to asserts.

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/test/java/io/opentracing/contrib/spring/web/servlet/TracingHandlerInterceptorTest.java:
```diff
@@ -0,0 +1,256 @@
+
+    @Test
+    public void testAsyncDeferred() throws Exception {
+        {
```
Ditto

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/test/java/io/opentracing/contrib/spring/web/servlet/TracingHandlerInterceptorTest.java:
```diff
@@ -0,0 +1,256 @@
+
+    @Test
+    public void testControllerException() throws Exception {
+        {
```
Same here

pavolloffay commented:
@bensigelman @yurishkuro could you please review?

yurishkuro commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/SpanDecorator.java:
```diff
@@ -0,0 +1,106 @@
+                    URL url = new URL(httpServletRequest.getRequestURL().toString());
+                    String errorUrl = String.format("%s://%s:%s%s", url.getProtocol(), url.getHost(), url.getPort(),
+                            httpServletRequest.getAttribute(WebUtils.ERROR_REQUEST_URI_ATTRIBUTE));
+                    Tags.HTTP_URL.set(span, errorUrl);
```
I am not familiar with this part of the framework - why do you need to override the URL tag if it was already set in `onPreHandle`? Perhaps add a comment to this section.

yurishkuro commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/TracingHandlerInterceptor.java:
```diff
@@ -0,0 +1,102 @@
+/**
+ * Tracing handler interceptor for spring web. It creates a new span for an incoming request.
+ * Created span can be accessed via {@link HttpServletRequest#getAttribute(String)} with name
+ * {@link TracingHandlerInterceptor#ACTIVE_SPAN}.
```
This seems like it would severely reduce the usefulness of the instrumentation, since the span is not propagated to any outbound calls this service might make. Basically, the instrumentation like this is most useful when there's an established standard for in-process context/span propagation.

pavolloffay replied:
If you want to make outbound calls you need to active span from attributes and pass it to instrumented client lib, which is not very handy at the moment.

As it was mentioned on gitter channel I hope that there will be some API changes to access active span context.

pavolloffay replied:
> Personally, I would do this differently. I would log the following fields only:
> event: "exception"
> exception: throwable
> and let the tracer implementation deal with exact representation of the exception / stack trace.

I agree this depends on tracer impl, probably users using some specific tracer would use different  decorator. As this is a "standard"  decorator, I wanted to add the most common tag/event to represent an error in the system.

pavolloffay replied:
This is based on    https://github.com/opentracing/specification/issues/29#issuecomment-270962980

bensigelman commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/SpanDecorator.java:
```diff
@@ -0,0 +1,107 @@
+            if (ex != null) {
+                Tags.ERROR.set(span, Boolean.TRUE);
+                span.log(logsForException(ex));
+            } else if (httpServletRequest.getAttribute(WebUtils.ERROR_EXCEPTION_ATTRIBUTE) != null) {
```
I think a 5xx status code is also grounds for `Tags.ERROR`...

bensigelman commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/TracingHandlerInterceptor.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    @Override
+    public void afterCompletion(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse,
```
so `afterCompletion` is the only hook we can use here? it's too bad that it gets called twice sometimes... the SpanWrapper feels a little gross.

pavolloffay replied:
IIRC `postHandle` could be used, however it does not expose exception... 

I will dig a little bit more and see if there is a cleaner way.

pavolloffay replied:
I raised an issue in spring jira: https://jira.spring.io/browse/SPR-15217. Hopefully it will be resolved.

bensigelman replied:
I assume you'd like to merge this in the meantime?

pavolloffay replied:
Spring boot has a generic error handling and after "normal" processing it invokes `preHandler` and `afterCompletion` for error handling/view `/error`. I thought that it invokes only `afterCompletion`...

I see two possible solutions:
1. use span wrapper with idempotent finish
2. remove span from request attributes in `afterCompletion`. Then it would start new span for `/error`.
 
@bensigelman what do you think?

pavolloffay commented:
Update:

I have added the same test for Spring MVC deployed on jetty and following scenarios interceptor in not invoked:

* an exception in filter
* request which does not match any URL 
* unauthenticated requests

In spring boot the interceptor is invoked in all of these use cases. I will do a bit more research and if not I will document it and refer to java-web-filter.

mabn commented:
@pavolloffay 

You can use:
org.springframework.web.context.request.RequestAttributes
org.springframework.web.context.request.ServletRequestAttributes

to check if currently an error is being handled. I'm not sure how reliable this is. It looks like this (when used inside error controller implementation):
```
    public String stuff(HttpServletRequest aRequest, HttpServletResponse response) {
        Map<String, Object> body = getErrorAttributes(aRequest, getTraceParameter(aRequest));
        // see DefaultErrorAttributes for list of attributes
        Integer status = (Integer) body.get("status");
        String error = (String) body.get("error");
        ...
    }

    private boolean getTraceParameter(HttpServletRequest request) {
        String parameter = request.getParameter("trace");
        if (parameter == null) {
            return false;
        }
        return !"false".equals(parameter.toLowerCase());
    }

    private Map<String, Object> getErrorAttributes(HttpServletRequest aRequest, boolean includeStackTrace) {
        RequestAttributes requestAttributes = new ServletRequestAttributes(aRequest);
        return errorAttributes.getErrorAttributes(requestAttributes, includeStackTrace);
    }
```

pavolloffay commented:
@mabn thanks, I know when an error is being handled.

I'm still not very happy with this... I will force consumers of this instrumentation to use web filter and this tracing interceptor together. That is the only way how to trace all requests for spring MVC (and boot). Span will be started in filter and interceptor will add spring related tags (if any).

bensigelman commented:
@pavolloffay I have been reading the javadocs for the Spring stuff. How infuriating that the `afterCompletion` handler gets called in these non-uniform ways! Not very accurately-documented, either.

One idea would be to use `FollowsFrom` here... create a Span for the time between the `pre` call and the *first* `afterCompletion` call; then create additional `FollowsFrom` Spans for any additional `afterCompletion` calls that are errors. This is basically what you suggested above in one of the comment threads, though you didn't specify `FollowsFrom` :)

This would capture state about what's going on. Whether the tracing system will display it sanely is another matter. But in some way I think I prefer this to hacks that rely on the web filter, etc.

My two cents. A bit of a predicament :-/

(A nuclear option would be a PR against Spring itself, as this seems busted to me... but that's a long-term fix that doesn't address the software being used in the wild.)

mabn commented:
> this to hacks that rely on the web filter

I wouldn't consider it a hack (the code I pasted is a hack though). Spring MVC is based on servlets and lives inside a DispatcherServlet. Spring Security adds a filter various authentication mechanisms. The most reasonable way to cover all these behaviours is to add tracing servlet filter early in the filter chain and then customize current span (alter operation_name, add tags...) inside other filters / interceptors.

Tracing error controlled execution as FollowsFrom would be kind of weird because the parent span (http request) has to wait for error controller execution anyway.

pavolloffay commented:
(btw. spring cloud sleuth uses servlet filter with interceptor, I think it is the only way how to properly get trace data for all use cases [no URL mapping, unauthenticated requests..], see https://jira.spring.io/browse/SPR-15226) 

The thing is `preHandler` and `afterCompletion` are always called only once(*) per handler invocation.  When you in `foo` handler  return `"forward:bar"`  the result is:

1. preHande `/foo`
2. preHandle `/bar`
3. afterCompletion `/bar`
4. afterCompletion `/foo`

and  I think this can happen often. My issue with an exception in a cotroller (deployed in embedded boot) is that request is just forwarded e.g. `forward:error` and as mentioned above it is a valid use case (and spring boot does this by default for errors).

I think we should add a log or create new span in every invocation of `preHandle`. Now it is a modeling decision:

When used filter+interceptor:
1. filter creates root span.
XX interceptor add logs in `preHandler` and `afterCompletion`
YY interceptor creates a child span of the span created in the filter or of a span on stack. Created span is added on stack. Span is finished and removed from the stack in `afterCompletion`.
2. filter finishes root span.

HTTP tags and errors are added in filter.

XX results in one span per request. In a handler  you can create a local span representing operation in a given handler, but to find out in which handler it was created you need to look at logs. If we go with YY then there would be one root span for a request and N children for each handler invocation and if you create a span in a handler you know more explicitly in which handler it happened. YY gives you a cleaner span structure but is more verbose.

When used without filter:
1. interceptor creates span in preHandle, if there is something on stack it whould be its parent. 
2. Span is finished and removed from stack in `afterCompletion`.

HTTP tags would be added in interceptor decorator. Decorator would also know #number of created spans before (to do not add the same tags again). 

 (*)Spring MVC execute sequence for async:

```
preHandle
afterConcurrentHandlingStarted
preHandle
postHandle
afterCompletion
```

in spring boot it's just `preHandle` and `afterCompletion`. This is not very nice because I would have to somehow skip the second `preHandle`. Use span wrapper: add some indication in `afterConcurrent` that next preHandle should be skipped.

Based on all this I think we should go with filter+interceptor together and then decide if a child span should be created or not.

@mabn @bensigelman @yurishkuro

pavolloffay commented:
Flow for Spring boot deployed in embedded container when an exception is thrown in a controller:

1. preHandle `/controller`
2. afterCompletion `/controller`
XX**. Exception caught in filter and span is finished**
3. preHandler for `/error`
4. afterCompletion `/error`

Based on this sometimes I would have to start span in interceptor and have a way to check if parent span from filter is finished or not... :worried:

bensigelman commented:
@pavolloffay I do not know enough about Spring to provide very useful feedback here. If it would be useful to get on a google hangout or similar to brainstorm possible workarounds, I'm happy to do so, for what it's worth.

pavolloffay commented:
@bensigelman @mabn I have updated the PR description. See how does it work. The goal is that it maximizes the use of span created in filter and creates additional span only if necessary (e.g. when span from filter is finished). 

I'm happy to see

objectiser commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/interceptor/SpanDecorator.java:
```diff
@@ -0,0 +1,114 @@
+                                      Object handler, Exception ex, Span span) {
+            Map<String, Object> logs = new HashMap<>(2);
+            logs.put("event", "afterCompletion");
+            logs.put("handler", handler);
```
Shouldn't this be same as `onPreHandler` - use `MetaDataUtils.HANDLER`, but also log the class and method names separately?

objectiser commented on opentracing-spring-web-itest/common/src/main/java/io/opentracing/contrib/spring/web/interceptor/itest/common/AbstractBaseITests.java:
```diff
@@ -0,0 +1,436 @@
+
+        span = mockSpans.get(1);
+        Assert.assertEquals(0, span.tags().size());
+        assertLogEvents(span.logEntries(), Arrays.asList("preHandle", "afterCompletion"));
```
May be worth checking that this span is a child of the server span?

objectiser commented on opentracing-spring-web-itest/common/src/main/java/io/opentracing/contrib/spring/web/interceptor/itest/common/AbstractBaseITests.java:
```diff
@@ -0,0 +1,436 @@
+        }
+    }
+
+    public static void assertErrors(List<MockSpan> spans) {
```
Would be better named `assertNoErrors`

objectiser commented on opentracing-spring-web-itest/common/src/main/java/io/opentracing/mock/TextMapPropagator.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * @author Pavol Loffay
+ */
+public class TextMapPropagator implements MockTracer.Propagator {
```
Is this here while waiting for release of opentracing-java? If so, 0.20.8 is now available

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/interceptor/InterceptorSpanWrapper.java:
```diff
@@ -0,0 +1,26 @@
+ */
+class InterceptorSpanWrapper extends SpanWrapper {
+
+    private Object handler;
```
What is this handler? Can't you add a parameter to this class, so that you have a real type, instead of using `Object`?

pavolloffay replied:
It is a handler object passed to the spring interceptor, they use object too.

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/interceptor/InterceptorSpanWrapper.java:
```diff
@@ -0,0 +1,26 @@
+import io.opentracing.contrib.web.servlet.filter.SpanWrapper;
+
+/**
+ * @author Pavol Loffay
```
Would be nice to have a line or two that explains that this class is about.

pavolloffay replied:
+1 Will write something, however it is an internal class.

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/interceptor/TracingHandlerInterceptor.java:
```diff
@@ -0,0 +1,147 @@
+        }
+
+        for (SpanDecorator decorator: decorators) {
+            decorator.onPreHandle(httpServletRequest, handler, serverSpan);
```
What happens if a decorator fails? Will it propagate to the request? Should it?

pavolloffay replied:
Good comment, I'm assuming that decorators won't throw any exceptions.

jpkrohling commented on opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/interceptor/TracingHandlerInterceptor.java:
```diff
@@ -0,0 +1,147 @@
+        }
+
+        for (SpanDecorator decorator: decorators) {
+            decorator.onAfterCompletion(httpServletRequest, httpServletResponse, handler, ex, serverSpan);
```
Same here: what happens if a decorator throws an exception? Should it propagate to the end user?

jpkrohling commented on opentracing-spring-web-itest/boot/src/test/java/io/opentracing/contrib/spring/web/interceptor/itest/boot/BootITest.java:
```diff
@@ -0,0 +1,36 @@
+ * @author Pavol Loffay
+ */
+@SpringBootTest(
+        webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT,
```
\o/

jpkrohling commented on opentracing-spring-web-itest/common/src/main/java/io/opentracing/contrib/spring/web/interceptor/itest/common/app/TestController.java:
```diff
@@ -0,0 +1,125 @@
+    public Callable<String> async() {
+        return new Callable<String>() {
+            public String call() throws Exception {
+                Thread.sleep(1000);
```
We talked about a similar code in a previous PR, and I have similar concerns here :)

jpkrohling commented on opentracing-spring-web-itest/mvc/src/test/java/io/opentracing/contrib/spring/web/interceptor/itest/mvc/SpringMVCConfiguration.java:
```diff
@@ -0,0 +1,81 @@
+    private List<SpanDecorator> spanDecorators;
+
+    @Autowired
+    private io.opentracing.Tracer tracer;
```
Why are you using the fully qualified class name here?

jpkrohling commented on opentracing-spring-web-itest/mvc/src/test/webapp/WEB-INF/jsp/staticView.jsp:
```diff
@@ -0,0 +1,3 @@
+<!DOCTYPE html>
+<body/>
```
What is this file for? Is this something that Spring requires?

pavolloffay replied:
It is used to test `ControllerView` and returning view.

jpkrohling commented on pom.xml:
```diff
@@ -0,0 +1,146 @@
+
+    <version.com.jayway.awaitility-awaitility>1.7.0</version.com.jayway.awaitility-awaitility>
+    <version.io.opentracing>0.20.9</version.io.opentracing>
+    <version.io.opentracing.contrib-opentracing-web-servlet-filter>0.0.3-SNAPSHOT</version.io.opentracing.contrib-opentracing-web-servlet-filter>
```
Just to confirm: you'll change this before merging, right?

pavolloffay replied:
Yep, the filter needs to be released. (CI is failing)

pavolloffay commented on README.md:
```diff
@@ -0,0 +1,72 @@
+* exceptions in filters
+
+, however these requests are not traced in Spring MVC application deployed in tomcat/jetty...
+To solve this problem consider using [web filter instrumentation](https://github.com/pavolloffay/java-spring-web).
```
Thanks, the whole Readme has to be updated.

bensigelman commented:
@pavolloffay basic question: I would have casually expected that exceptions thrown during servlet/interceptor invocation would lead to `"error"=true` tags, but I don't see that in the PR – apologies if I missed something relevant in this PR thread, but why is that?

pavolloffay commented:
These errors are propagated to the web filter. Web filter is basically adding all tags (HTTP, error). You can see it in tests that error tag/logs are present.

bensigelman commented on opentracing-spring-web-itest/mvc/src/test/java/io/opentracing/contrib/spring/web/interceptor/itest/mvc/MVCJettyITest.java:
```diff
@@ -0,0 +1,197 @@
+        Assert.assertEquals(getUrl("/exception"), span.tags().get(Tags.HTTP_URL.getKey()));
+        Assert.assertEquals(500, span.tags().get(Tags.HTTP_STATUS.getKey()));
+        Assert.assertNotNull(span.tags().get(Tags.COMPONENT.getKey()));
+        Assert.assertEquals(Boolean.TRUE, span.tags().get(Tags.ERROR.getKey()));
```
Yes, I see this now. Thanks @pavolloffay. (I was previously looking for non-test code that sets the error tag)

pavolloffay commented:
@mabn @jpkrohling @objectiser @yurishkuro @bensigelman thanks for review/s :+1: !

pavolloffay commented:
Issue created: https://jira.spring.io/browse/SPR-15265
